Adds Selectize support for array and select editors

philostler opened the pull request:
PR by myself @philostler, @radioisradio and @JulianKigwana to add Selectize support to json-editor.

Before:
![json_editor_example](https://cloud.githubusercontent.com/assets/244198/7707393/6f566968-fe47-11e4-915c-a88768e389c9.png)
After:
![json_editor_example 2](https://cloud.githubusercontent.com/assets/244198/7707400/76cce14a-fe47-11e4-8377-c7dfe3aea7dd.png)
And a new Selectize example:
![json_editor_selectize_integration_example](https://cloud.githubusercontent.com/assets/244198/7707406/818bb17e-fe47-11e4-94fc-67611bc21b38.png)

**Notes:**
We've made every effort to remain backwards compatible with how json-editor works as of now. So, as before, Select2 will be used if it is present (we didn't want to break this default behaviour).

If you want Selectize turned on then the following snippet does that for you:
```js
JSONEditor.plugins.selectize.enable = true;
```

I think, given that you can now use Select2 and Selectize, there's a case for seperating the editors out which is what we began thinking about. You'll see that in that `array` editor for Selectize is in it's own module. 

Ultimately though that would be a larger change (and a major/minor revision increment) that would probably make selecting between Select2 and Selectize nicer as well. One idea would be to pass in the prefer enhanced UI in to the editor options. e.g.
```js
this.editor = new JSONEditor(element, {
    enhanced_ui: 'selectize', // or 'select2' or anything else for 'default'
    schema: schema
});
```

We didn't want to undertake that as well, so this PR only includes Selectize support and no new ideas on how to better select your enhanced UI options

Enjoy!

brettz9 commented:
Just a thought that maybe a "format" property in conjunction with `uniqueItems`, could also trigger tag selection (and also suggest that the ordering of the array was not significant).

philostler commented:
@brettz9 There of course many other scenarios where it would be extremely helpful to use the tag selection (enums using a limited allowed selection could be one).

We've not covered all those scenarios, but made a start on the basics. Maybe something to put in on a roadmap for the library (is there is one)

philostler commented:
@jdorn Any thoughts on this?

jdorn commented on src/core.js:
```diff
@@ -177,6 +177,9 @@ JSONEditor.prototype = {
     if(!classname) throw "Unknown editor for schema "+JSON.stringify(schema);
+    if(classname === "array" && JSONEditor.plugins.selectize.enable) {
+        classname = classname + "Selectize";
+    }
```
Can this be done with the resolver functions instead?  I don't like having this logic in core.js

philostler replied:
Yep, it indeed can be done with resolvers! I've removed this now and moved that logic into a resolver of it's own

jdorn commented on src/editors/array.js:
```diff
@@ -32,11 +32,11 @@ JSONEditor.defaults.editors.array = JSONEditor.AbstractEditor.extend({
     if(this.remove_all_rows_button) this.remove_all_rows_button.disabled = false;
     if(this.delete_last_row_button) this.delete_last_row_button.disabled = false;
-    
+
```
It looks like the only thing that changed in this whole file was whitespace.  Can you revert this file to keep the pull request clean?

philostler replied:
Reverted

jdorn commented on src/editors/select.js:
```diff
@@ -318,4 +318,4 @@ JSONEditor.defaults.editors.select = JSONEditor.AbstractEditor.extend({
     this._super();
   }
-});
+});
```
Can you revert this file?  It's just a whitespace change and I want to keep the pull request as small as possible to make it easier to merge.

philostler replied:
Reverted

jdorn commented on src/themes/bootstrap3.js:
```diff
@@ -44,9 +44,8 @@ JSONEditor.defaults.themes.bootstrap3 = JSONEditor.AbstractTheme.extend({
-    else {
-      group.className += ' form-group';
+    } else {
+      group.className += 'form-group';
```
Can you revert this file? It's just formatting and doesn't seem relevant to the pull request.

philostler replied:
Reverted

jdorn commented on demo.html:
```diff
@@ -10,7 +10,10 @@
+    <script src="//cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
+    <script src="//cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.1/js/standalone/selectize.js"></script>
+    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.1/css/selectize.bootstrap3.css">
+
```
Can these be loaded only when selectize support is enabled?  I don't like having blocking calls in the head. A similar thing is done for the theme and icon css files.

philostler replied:
I've make these load on demand now. See [here](https://github.com/philostler/json-editor/commit/05d25153d505877103a5da7688fa74d958b3b055)

jdorn commented:
Sorry for taking so long to look at this PR. It looks good for the most part, but I added some minor comments for things I'd like changed before merging.  Thanks!

philostler commented:
@jdorn Sorry it's taken a while to go back and refactor this with your comments, busy busy.

I've taken on board everything you commented on and addressed all your points of concern. Please let me know if there is anything I've missed.

I've of course brought this PR back up-to-date with master and additionally I've improved the Selectize code so commas can now be added to the input.

Please review and let me know if there is anything else I can do to get this PR merged!
